colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,290 @@
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+    
+    # approx_sq_grad for row parallel weight
+    @staticmethod
+    def _approx_sq_grad_row_parallel(exp_avg_sq_row, exp_avg_sq_col, sq_row_meam):
+        # row_meam = sq_row_meam
+        r_factor = (exp_avg_sq_row / sq_row_meam).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+        
+    @torch.no_grad()
+    def step(self, closure=None):
+        """
+        Performs a single optimization steps
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+        """
+        param_groups: Dict
+        {
+            "params":[weight, bias]
+            "lr"
+            "eps"
+            "clip_threshold"
+            "decay_rate"
+            "beta1"
+            "weight_decay"
+            "scale_parameter"
+            "relative_step"
+            "warmup_init"
+        }
+        """
+        # print(f"param_groups {self.param_groups}")
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                
+                # ==============================
+                # Init Current Param Shape
+                # ==============================
+                p_shape = None  # param shape (not split)
+                param_height = None
+                param_width = None
+                param_height_parallel = None
+                param_width_parallel = None
+                sharding_spec = None
+                if id(p) in self.param_shape.keys():
+                    p_shape = self.param_shape[id(p)]
+                    # if id(p) in self.sharding_spec_dict.keys() and self.sharding_spec_dict[id(p)] != None:
+                    #     sharding_spec = self.sharding_spec_dict[id(p)]
+                    if len(p_shape) >= 2: # factored
+                        param_height = p_shape[0]
+                        param_width = p_shape[1]
+                        sharding_spec = self.sharding_spec_dict[id(p)]
+                        if sharding_spec.sharding_sequence[0] == 'R': # Col Parallel 
+                            param_height_parallel = param_height
+                            param_width_parallel = param_width // self.tensor_parallel_size # W/N
+                        if sharding_spec.sharding_sequence[-1] == 'R': # Row Parallel
+                            param_height_parallel = param_height // self.tensor_parallel_size # H/N
+                            param_width_parallel = param_width
+                
+                # grad shape is same as weigh / bias
+                grad = p.grad.to(self.localRank)
```

Comment:
Don't move grad here

Reply:
> Don't move grad here

fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/d8eae847049e6062454257379caf7abaae091756